File: bulldog/services/expense_service.go
package services

import (
	"github.com/greysespinoza/fs-path/models"
	"github.com/greysespinoza/fs-path/repositories"
	"time"
	"errors"
)

type ExpenseService struct {
	Repo *repositories.ExpenseRepository
}

func NewExpenseService(repo *repositories.ExpenseRepository) *ExpenseService {
	return &ExpenseService{Repo: repo}
}

func (s *ExpenseService) CreateExpense(exp *models.Expense) error {
	if exp.UserID == 0 || exp.Amount <= 0 || exp.Category == "" || exp.Date.IsZero() {
		return errors.New("all fields are required and amount must be positive")
	}
	return s.Repo.CreateExpense(exp)
}

func (s *ExpenseService) GetExpenses() ([]models.Expense, error) {
	return s.Repo.GetExpenses()
}

func (s *ExpenseService) GetExpenseByID(id int) (*models.Expense, error) {
	return s.Repo.GetExpenseByID(id)
}

func (s *ExpenseService) UpdateExpense(exp *models.Expense) error {
	// Validate fields
	if exp.UserID == 0 || exp.Amount <= 0 || exp.Category == "" || exp.Date.IsZero() {
		return errors.New("all fields are required and amount must be positive")
	}

	_, err := time.Parse("2006-01-02", exp.Date.Format("2006-01-02"))
	if err != nil {
		return errors.New("invalid date format")
	}

	return s.Repo.UpdateExpense(exp)
}

func (s *ExpenseService) DeleteExpense(id int) error {
	return s.Repo.DeleteExpense(id)
}

File: bulldog/services/expense_service.py
from repositories.expense_repository import ExpenseRepository
from repositories.bank_account_repository import BankAccountRepository

class ExpenseService:
    @staticmethod
    def create_expense(data):
        if not data.get("bank_account_id") or not data.get("amount") or not data.get("category") or not data.get("date") or not data.get("hour"):
            raise ValueError("All fields (user_id, amount, category, date, hour) are required.")
        expense = ExpenseRepository.create_expense(data)
        from decimal import Decimal
        updated_balance = expense.bankAccount.balance - Decimal(str(data['amount']))
        BankAccountRepository.update_bank_account(expense.bank_account_id, {'balance': updated_balance})

    @staticmethod
    def get_expenses(page, per_page):
        return ExpenseRepository.get_expenses(page, per_page)

    @staticmethod
    def get_expense_by_id(expense_id):
        return ExpenseRepository.get_expense_by_id(expense_id)

    @staticmethod
    def get_expenses_by_user_id(user_id):
        return ExpenseRepository.get_expenses_by_user_id(user_id)

    @staticmethod
    def update_expense(expense_id, data):
        if not data.get("user_id") or not data.get("amount") or not data.get("category") or not data.get("date") or not data.get("hour"):
            raise ValueError("All fields (user_id, amount, category, date, hour) are required.")
        ExpenseRepository.update_expense(expense_id, data)

    @staticmethod
    def delete_expense(expense_id):
        ExpenseRepository.delete_expense(expense_id)
